https://issues.redhat.com/browse/ACM-23758 forwarding alerts mcoa (#8428)
open-cluster-management/rhacm-docs: observability/obs_alert_forward_mcoa.adoc

[#forward-alert-mcoa]
= Forwarding alerts for multicluster observability add-on

Configure alert forwarding for the multicluster observability add-on to send alerts to the Alertmanager of the Observability stack from your managed clusters. By default, multicluster observability add-on is not configured to send alerts from Prometheus on your managed clusters to the hub cluster Alertmanager.

> Create a policy to configure alert forwarding for the multicluster observability add-on.

Reviewer: ```suggestion
To setup alert forwarding when using the multicluster observability add-on, you can create a policy.
```

Would prefer it like this, but it's a small detail, and maybe it better matches guidelines as is, so I'll leave it to you to decide.

Reviewer 2: The way it is written, yes it matches the guidelines for conciseness and imperative language. The point is to start with the action first and not start with a conditional phrase or any other phrase. 

Your way is not wrong, we are just encouraged to lead with the action first. @jacobbaungard -- so I am not against it 100%--it's just not leading with the verb "Create," which is the point of the step.

Does this work?

_Create a policy to configure alert forwarding when you use the multicluster observability add-on._

Reviewer: Yes that is also fine :+1:

                  insecureSkipVerify: false
              externalLabels:
                managed_cluster: %[3]s
          ` }}

            {{- $mangedConfig = merge $mangedConfig
                              ((printf $pm $hubName $hubBaseDomain (fromClusterClaim "id.openshift.io"))| fromYaml)
          }}
          - complianceType: mustonlyhave
            objectDefinition:
              apiVersion: v1
              data:
                config.yaml: |
                  {{ (merge $cy $mangedConfig)| toYaml |autoindent }}
              kind: ConfigMap
              metadata:
                name: cluster-monitoring-config
                namespace: openshift-monitoring
            recordDiff: InStatus
  remediationAction: inform
----

+
*Note:* The `remediationAction` of the previous policy is set to `inform` mode. From the console, validate your changes by viewing the policy from the *Governance* page. Your policy appears as non-compliant.

.. To apply the policy changes across all managed clusters and enable alert forwarding from your managed clusters to the hub cluster, change the `remediationAction` parameter value to `enforce`. 

> .. Create a `PlacementBinding` resource to bind the policy to the `global` placement. Your YAML file might resemble the following sample:

Reviewer: I'm a bit unsure, if it's better to move the placement binding steps, directly above in L92.

The reason being:
- We add the content to the existing file (maybe we can clarify that also)
- The actions the note on line L94 will only happen after both the policy and the placementbinding is applied

Would that make sense ?

Author: Yea that makes sense. I will move the note. it will be after the PlacementBinding step

[#forward-alert-mcoa]

> = Forwarding alerts for multicluster observability add-on

Reviewer 2: I had this question on the others:

multicluster observability add-on (is this coded?) If not, this isn't a title?

Author: When you say coded are you referring to it being added to the conref file?

Reviewer 2: no, is the add on name supposed to be coded in the titles and such...I am not sure bc I didn't review this work.

Author: no it shouldn't be coded or monospaced

          }}
          - complianceType: mustonlyhave
            objectDefinition:
              apiVersion: v1
              data:
                config.yaml: |
                  {{ (merge $cy $mangedConfig)| toYaml |autoindent }}
              kind: ConfigMap
              metadata:
                name: cluster-monitoring-config
                namespace: openshift-monitoring
            recordDiff: InStatus
  remediationAction: inforce
----
//you mention `inform` mode, but I only see `enforce` for the remediation action. Please clarify. Do we need to mention anything about `inform` mode here?
.. Create a `PlacementBinding` resource to bind the policy to the `global` placement. Your YAML file might resemble the following sample:

+
[source,yaml]
----
apiVersion: policy.open-cluster-management.io/v1
kind: PlacementBinding
metadata:
  name: mcoa-alert-forward-platform-placement
  namespace: open-cluster-management-global-set
placementRef:
  apiGroup: cluster.open-cluster-management.io
  kind: Placement
  name: global
subjects:
- apiGroup: policy.open-cluster-management.io
  kind: Policy
  name: mcoa-alert-forward-platform
----

> .. Apply your YAML file. If you make custom changes to the `additionalAlertmanagerConfigs` specification of the config map, be sure to add the changes directly into the policy, before applying the policy. Run the following command:

Reviewer 2: there is a command to apply usually.

Author: There is a command on line 125-128. At first the following statement was a note: "If you make custom changes to the `additionalAlertmanagerConfigs` specification of the config map, be sure to add the changes directly into the policy, before applying the policy."

If the note is something the user needs to consider before running the command, it should be included in the step. Or i can just add as a note like before. let me know your thoughts

Reviewer: I think perhaps we should move this paragraph a bit. Either before the initial yaml at L22 or just after the yaml at L114

Reviewer: To clarify

> If you make custom changes to the `additionalAlertmanagerConfigs` specification of the config map, be sure to add the changes directly into the policy, before applying the policy.

Could be moved, but not the part about applying the yaml / running the command.

Reviewer 2: The step itself is "apply" and the user needs to know how to do that right away as part of the step. The info is important and can be presented before or after the step. But what is the actual step here? It is "Apply".

          {{- $cy := dict }}
          {{- if and $cmo $cmo.data }}
            {{- $cy = (index $cmo "data" "config.yaml") | fromYaml }}
          {{- end }}
----

.. Define what you want your configuration to include. Your YAML file might resemble the following partial example:

+
[source,yaml]
----
          {{- $mangedConfig := dict }}
 
          {{- $pm := `
            prometheus:
              additionalAlertmanagerConfigs:
              - apiVersion: v2
                bearerToken:
                  key: token
                  name: observability-alertmanager-accessor-%[1]s
                scheme: https
                staticConfigs:
                - alertmanager-open-cluster-management-observability.apps.%[2]s
                tlsConfig:
                  ca:
                    key: service-ca.crt
                    name: hub-alertmanager-router-ca-%[1]s
                  insecureSkipVerify: false
              externalLabels:
                managed_cluster: %[3]s
          ` }}
----

.. Merge all of the configurations into one. Your YAML file might resemble the following partial example:

> 

Reviewer 2: The step is to merge? What do we mean by this? What is the reason for breaking this part out?

Author: It was a hidden comment in the original issue. The YAML is really long so I pulled explanation and used it as the purpose for specific parts of the YAML

Reviewer 2: Long YAML is OK if it doesn't require many steps/procedures in the YAML. We just need to pull out what requires a procedure:

Example:

. Add the ___ specification for the ___ to your __ resource. Change the ___ value to your specific namespace. See the following YAML sample:

```
insert that piece of the yaml
```
. Run the following command to apply your YAML file.

```
oc apply ___
```

This is just an example of how/why we would break out a large YAML sample to build a procedure.
